dags/sql/select-find_gaps.sql

```diff
@@ -0,0 +1,77 @@
+--a generic find_gaps function to find gaps in sensor data
+--includes a lookback of `gap_threshold` so it catches gaps overlapping the start of the interval
+--includes synthetic start/end time points for all sensors so we catch gaps that don't start or end in the interval
+    --doesn't catch sensors with no data in the interval. 
+--noqa: disable=TMP, PRS
+
+WITH raw AS (
+    SELECT DISTINCT
+        {{ params.id_col }} AS sensor_id_col,
+        {{ params.dt_col }} AS dt_col
+    FROM {{ params.table }}
+    WHERE
+        {{ params.dt_col }} >= '{{ ds }} 00:00:00'::timestamp - '{{ params.gap_threshold }}'::interval -- eg. >= 23:00
+        AND {{ params.dt_col }} < '{{ ds }} 00:00:00'::timestamp + '1 day'::interval
+),
+
+--add a start and end dt for each ID which appears in raw data
+fluffed_data AS (
+    --synthetic data
+    SELECT DISTINCT
+        raw.sensor_id_col,
+        bins.dt_col,
+        --don't need to reduce gap width for synthetic data since that minute does not contain data
+        '0 minutes'::interval AS gap_adjustment
+    FROM raw, 
+        (VALUES 
+            ('{{ ds }} 00:00:00'::timestamp - '{{ params.gap_threshold }}'::interval),
+            ('{{ ds }} 00:00:00'::timestamp + '1 day'::interval)
+        ) AS bins(dt_col)
+    
+    UNION ALL
+    
+    SELECT
+        sensor_id_col,
+        dt_col,
+        --need to reduce gap length by 1 minute for real data since that minute contains data
+        '{{ params.default_bin }}'::interval AS gap_adjustment
```

Comment:
is there a use case for this to not be 1 min?

Reply:
For different volumes datasets with different default reporting intervals! Admittedly I haven't tested on anything except for the 1-minute miovision data.

Reply:
ahhhh make sense